lint: namedConst suggestions ignore import-with-rename

From etcd:
```
check-package: $GOPATH/src/github.com/coreos/etcd/contrib/recipes/barrier.go:39:36: namedConst: use clientv3.NoLease instead of 0
check-package: $GOPATH/src/github.com/coreos/etcd/contrib/recipes/key.go:50:41: namedConst: use clientv3.NoLease instead of 0
```
It does suggest `clientv3.NoLease`, but `clientv3` imported as `v3`, so it should say `v3.NoLease` instead.

--- a/lint/lint.go
+++ b/lint/lint.go
@@ -95,6 +95,9 @@ func (c *Checker) Check(f *ast.File) []Warning {
 	if c.ctx.require.pkgObjects {
 		resolvePkgObjects(c.ctx.Context, f)
 	}
+	if c.ctx.require.pkgRenames {
+		resolvePkgRenames(c.ctx.Context, f)
+	}
 	c.walker.WalkFile(f)
 	return c.ctx.warnings
 }
@@ -135,10 +138,16 @@ type Context struct {
 	// For example, Context.require.pkgObjects => Context.pkgObjects.
 	require struct {
 		pkgObjects bool
+		pkgRenames bool
 	}
 
 	// pkgObjects stores all imported packages and their local names.
 	pkgObjects map[*types.PkgName]string
+
+	// pkgRenames maps package path to its local renaming.
+	// Contains no entries for packages that were imported without
+	// explicit local names.
+	pkgRenames map[string]string
 }
 
 // See #614.

--- a/lint/lint_private.go
+++ b/lint/lint_private.go
@@ -208,3 +208,15 @@ func resolvePkgObjects(ctx *Context, f *ast.File) {
 		}
 	}
 }
+
+func resolvePkgRenames(ctx *Context, f *ast.File) {
+	ctx.pkgRenames = make(map[string]string)
+
+	for _, spec := range f.Imports {
+		if spec.Name != nil {
+			obj := ctx.typesInfo.ObjectOf(spec.Name)
+			path := obj.(*types.PkgName).Imported().Path()
+			ctx.pkgRenames[path] = spec.Name.Name
+		}
+	}
+}

--- a/lint/lint_private.go
+++ b/lint/lint_private.go
@@ -6,6 +6,7 @@ import (
 	"go/types"
 	"log"
 	"reflect"
+	"strconv"
 	"strings"
 
 	"github.com/go-critic/go-critic/lint/internal/astwalk"
@@ -214,8 +215,10 @@ func resolvePkgRenames(ctx *Context, f *ast.File) {
 
 	for _, spec := range f.Imports {
 		if spec.Name != nil {
-			obj := ctx.typesInfo.ObjectOf(spec.Name)
-			path := obj.(*types.PkgName).Imported().Path()
+			path, err := strconv.Unquote(spec.Path.Value)
+			if err != nil {
+				panic(err)
+			}
 			ctx.pkgRenames[path] = spec.Name.Name
 		}
 	}

--- a/lint/namedConst_checker.go
+++ b/lint/namedConst_checker.go
@@ -25,6 +25,10 @@ return pos != 0`
 	d.After = `return pos != token.NoPos`
 }
 
+func (c *namedConstChecker) Init() {
+	c.ctx.require.pkgRenames = true
+}
+
 func (c *namedConstChecker) EnterChilds(x ast.Node) bool {
 	switch x := x.(type) {
 	case *ast.BinaryExpr:
@@ -50,6 +54,13 @@ func (c *namedConstChecker) isTimeExpr(x *ast.BinaryExpr) bool {
 		return false
 	}
 	return astp.IsBasicLit(x.X) || astp.IsBasicLit(x.Y)
+}
+
+func (c *namedConstChecker) importedName(obj types.Object) string {
+	if renamed := c.ctx.pkgRenames[obj.Pkg().Path()]; renamed != "" {
+		return renamed
+	}
+	return obj.Pkg().Name()
 }
 
 func (c *namedConstChecker) VisitExpr(x ast.Expr) {
@@ -95,7 +106,7 @@ func (c *namedConstChecker) VisitExpr(x ast.Expr) {
 func (c *namedConstChecker) warn(cause ast.Node, v constant.Value, named *types.Const) {
 	suggestion := named.Name()
 	if named.Pkg() != c.ctx.pkg {
-		suggestion = named.Pkg().Name() + "." + suggestion
+		suggestion = c.importedName(named) + "." + suggestion
 	}
 	c.ctx.Warn(cause, "use %s instead of %s", suggestion, v)
 }

--- a/lint/testdata/namedConst/positive_tests.go
+++ b/lint/testdata/namedConst/positive_tests.go
@@ -1,7 +1,7 @@
 package checker_tests
 
 import (
-	"time"
+	mytime "time"
 )
 
 type color int
@@ -26,8 +26,8 @@ func rawReturn() {
 		return 0
 	}
 
-	_ = func() time.Duration {
-		/// use time.Nanosecond instead of 1
+	_ = func() mytime.Duration {
+		/// use mytime.Nanosecond instead of 1
 		return 1
 	}
 

